--- a/services/core/java/com/android/server/audio/AudioService.java
+++ b/services/core/java/com/android/server/audio/AudioService.java
@@ -289,16 +289,16 @@ public class AudioService extends IAudioService.Stub {
 
    /** Maximum volume index values for audio streams */
     private static int[] MAX_STREAM_VOLUME = new int[] {
-        5,  // STREAM_VOICE_CALL
-        7,  // STREAM_SYSTEM
-        7,  // STREAM_RING
-        15, // STREAM_MUSIC
-        7,  // STREAM_ALARM
-        7,  // STREAM_NOTIFICATION
-        15, // STREAM_BLUETOOTH_SCO
-        7,  // STREAM_SYSTEM_ENFORCED
-        15, // STREAM_DTMF
-        15  // STREAM_TTS
+        20,  // STREAM_VOICE_CALL
+        28,  // STREAM_SYSTEM
+        28,  // STREAM_RING
+        60, // STREAM_MUSIC
+        28,  // STREAM_ALARM
+        28,  // STREAM_NOTIFICATION
+        60, // STREAM_BLUETOOTH_SCO
+        28,  // STREAM_SYSTEM_ENFORCED
+        60, // STREAM_DTMF
+        60  // STREAM_TTS
     };
 
     /** Minimum volume index values for audio streams */
